rfcs/text/0015_nodejs_clustering.md

```diff
@@ -404,18 +404,65 @@ We will need to add some centralized status state in the coordinator. Also, as t
 from the coordinator, we will also need to have the workers retrieve the global status from the coordinator to serve 
 the status endpoint.
 
-We may also want to have the `/status` endpoint display each individual worker status in addition to the 
-global status, which may be breaking change in the `/status` API response format.
+Ultimately, we'd need to make the following updates to the `/status` API, neither of which
+is a breaking change:
+1. The response will return the highest-severity status level for each plugin, which will be
+determined by looking at the shared global status stored in the coordinator.
+2. We will introduce an extension to the existing `/status` response to allow inspecting
+per-worker statuses.
+
+### 6.1.4 The stats API & metrics service
+
+The `/stats` endpoint is somewhat problematic in that it contains a handful of `process` metrics
+which will differ from worker-to-worker:
+```json
+{
+  // ...
+  "process": {
+    "memory": {
+      "heap": {
+        "total_bytes": 533581824,
+        "used_bytes": 296297424,
+        "size_limit": 4345298944
+      },
+      "resident_set_size_bytes": 563625984
+    },
+    "pid": 52646,
+    "event_loop_delay": 0.22967800498008728,
+    "uptime_ms": 1706021.930404
+  },
+  // ...
+}
+```
+
+As each request could be routed to a different worker, different results may come back each time.
+
+This endpoint, registered from the `usage_collection` plugin, is getting these stats from Core's
+`metrics` service (`getOpsMetrics$`), which is also used in the `monitoring` plugin for stats
+collection.
 
-### 6.1.4 PID file
+While we could try to follow a plan similar to `/status` and have each worker report metrics to
+the coordinator, there isn't a particularly sensible way to consolidate these. (e.g. Taking
+the average of these values isn't really helpful in case there's a scenario where one process
+is, for example, experiencing a spike in `event_loop_delay` while the others are operating
+normally).
+
+Ultimately we could extend the API to provide per-worker stats, but the question remains what we
+should do with the existing `process` stats:
+1. Deprecate them? (breaking change)
+2. Accept a situation where they may be round-robined to different workers? (probably no)
+3. Try to consolidate them somehow? (can't think of a good way to do this)
+4. Always return stats for one process, e.g. main or coordinator? (doesn't give us the full picture)
```

Comment:
@joshdover @pgayvallet This is the last open question I've run into, would appreciate your thoughts on it.

I don't love any of these options as I'd really like to avoid a breaking change if we can help it, but maybe you'll have some other ideas 🙂

Reply:
I think we want to do the same than we're planning to do for the `/status` endpoint / data, and allow to access per-worker process stats.

Now, for the current existing `OpsMetrics.process`, this is an excellent question, and I don't have any obvious answer. I do think none of the `2.`, `3.` or `4.` option really makes sense, but `1.` is introducing a breaking change, which doesn't play well at all with our goals on this matter, as we're very unlikely to finish the implementation before `8.0`. So, between `2.`, `3.` and `4.`, I would say that `4.` is (maybe?) the lesser evil

Reply:
>  1. is introducing a breaking change, which doesn't play well at all with our goals on this matter, as we're very unlikely to finish the implementation before 8.0. So, between 2., 3. and 4., I would say that 4. is (maybe?) the lesser evil

These were my thoughts exactly, 4 feels like the least-bad option, but it's still not great.

> I think we want to do the same than we're planning to do for the /status endpoint / data, and allow to access per-worker process stats.

Yep, agreed on this too. I think we could solve for this in a variety of non-breaking ways. The ops metrics is the tricky part.

Reply:
Whenever we make this feature GA, we'll going need to have Metricbeat updated to be able to pull data for multiple nodes against a single endpoint and/or have some way to display multiple processes in the Stack Monitoring UI. If there are any Metricbeat changes needed, I think it would make sense to prioritize getting those changes in early in the project so that very few users are flying blind or with only partial data during the eventual upgrade to clustering by default. If we were able to get these changes in for 7.last, then we'd be sure that this data wouldn't break in monitoring.

But we still should try to make sure that super old Metricbeat instances get something sorta sensible. I think aggregating the data for all processes (option 3) is probably going to create less headache than any other option. Even if it's not 100% accurate or sensible in this format.

Here's what I would propose as a way to aggregate each field:

https://github.com/elastic/kibana/blob/master/src/core/server/metrics/collectors/types.ts#L21-L42

```ts
export interface OpsProcessMetrics {
  /** process memory usage */
  memory: {
    /** heap memory usage */
    heap: {
      /** total heap available */
      total_in_bytes: number; // sum
      /** used heap */
      used_in_bytes: number; // sum
      /** v8 heap size limit */
      size_limit: number; // sum
    };
    /** node rss */
    resident_set_size_in_bytes: number; // sum
  };
  /** node event loop delay */
  event_loop_delay: number; // max?
  /** pid of the kibana process */
  pid: number; // coordinator process
  /** uptime of the kibana process */
  uptime_in_millis: number; // coordinator process
}
```

**heap.size_limit** is probably the one that makes the least sense because it would be possible for the graph to be showing it hasn't hit the limit, but a process may actually be very close to approaching the limit. There is some discussion about removing the heap limit from the graph though: https://github.com/elastic/kibana/issues/100408

Reply:
@joshdover I think your proposal above makes sense:

- sum of available/used heap & node rss is straightforward
- `size_limit` sum could be confusing as you mention, but for the reasons explained in #100408, it's already creating confusion, so we aren't necessarily changing the status quo 🤷 
- `event_loop_delay` max makes sense to me. for debugging purposes we're mostly interested in that number if its high anyway. we could still discern from the logs which process has the high delay.
- `pid` and `uptime_in_millis` from the coordinator also make sense, especially with the initial implementation where we plan to kill all workers and coordinators if one of them dies. in the future if we respawn workers that die, this could be misleading, but hopefully by then we could have a solution in place on the Metricbeat side.

If we took this approach, the plan would then be:
- Use aggregated metrics in the initial implementation
- Add per-worker process metrics to the stats
  - Not sure exactly what this would look like -- ECS doesn't have a prescribed solution for this situation AFAICT -- but regardless we could do it in a non-breaking way
- Update Metricbeat module to use the new metrics format, and also Monitoring UI (hopefully by 7.last)

Long term we would still need to decide what to do about the existing `process`. Perhaps we could mark it as deprecated at least, even if we don't remove it in `8.0`. It wouldn't make sense to keep it around forever if the numbers could be misleading.

Reply:
>  Perhaps we could mark it as deprecated at least, even if we don't remove it in 8.0. 

I find the option with aggregation to be a bit confusing for the end-users since they cannot grasp the state of the system at a single glance. 

My +1 to deprecate it, but keep it for some time. We can even consider outputting the process data in a different format for single-(`process: OpsProcessMetrics`) and multi-clustering modes (`process: OpsProcessMetrics[]`). Or use another end-point for `/stats` in multi-clustering mode.

Reply:
I updated the language to clarify the plan: 

1. Deprecate `process` field & introduce aggregated metrics as an interim solution
2. Then later make the breaking changes, either removing `process` or supporting multiple formats as mentioned above

The good news is that we only really need step (1) to ship a first iteration of this, so we have some time to decide on the details around (2)